File: JavaScript/src/Internals/Functions/Financial/Function_IRR.js
import { Function_N } from '../Function_N.js';
import { Operand } from '../../../Operand.js';

class Function_IRR extends Function_N {
    get Name() {
        return "IRR";
    }

    constructor(funcs) {
        super(funcs);
    }

    evaluate(engine, tempParameter) {
        if (this.z.length < 1) return this.parameterError(1);

        const valuesArg = this.getArray(engine, tempParameter, 0);
        if (valuesArg.IsError) return valuesArg;
        const values = [];
        for (const v of valuesArg.ArrayValue) {
            if (v.IsNumber) {
                values.push(v.NumberValue);
            } else {
                const v2 = v.ToNumber(`Function '${this.Name}' parameter 1 is error!`);
                if (v2.IsError) return v2;
                values.push(v2.NumberValue);
            }
        }

        if (values.length === 0) return this.parameterError(1);

        // IRR 要求现金流量必须同时包含正负值
        let hasPositive = false;
        let hasNegative = false;
        for (const v of values) {
            if (v > 0) hasPositive = true;
            if (v < 0) hasNegative = true;
        }
        if (!hasPositive || !hasNegative) return this.parameterError(1);

        let guess = 0.1;
        if (this.z.length > 1) {
            const guessArg = this.getNumber(engine, tempParameter, 1);
            if (guessArg.IsError) return guessArg;
            guess = guessArg.NumberValue;
        }

        const irr = this.newtonRaphsonIRR(values, guess);
        if (irr === null) return this.functionError();
        return Operand.Create(irr);
    }

    newtonRaphsonIRR(values, guess) {
        let rate = guess;
        for (let iter = 0; iter < 100; iter++) {
            let npv = 0;
            let dnpv = 0;
            const onePlusRate = 1 + rate;
            let factor = 1;

            for (let i = 0; i < values.length; i++) {
                npv += values[i] / factor;
                dnpv -= i * values[i] / (factor * onePlusRate);
                factor *= onePlusRate;
            }

            if (Math.abs(dnpv) < 1e-12) break;
            const newRate = rate - npv / dnpv;

            if (Math.abs(newRate - rate) < 1e-10) {
                return newRate;
            }
            rate = newRate;
        }
        return null; // 未收敛
    }
}

export { Function_IRR };

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/functions/financial/Function_IRR.java
package toolgood.algorithm.internals.functions.financial;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;
import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.NoneEngine;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_2;

final class Function_IRR extends Function_2 {
    public Function_IRR(FunctionBase[] funcs) {
        super(funcs);
        if (funcs.length < 1 || funcs.length > 2) {
            throw new IllegalArgumentException(String.format("Function '%s' requires 1 to 2 parameters.", Name()));
        }
    }

    @Override
    public String Name() {
        return "IRR";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand valuesArg = GetArray_1(engine, tempParameter);
        if (valuesArg.IsErrorOrNone()) return valuesArg;
        List<BigDecimal> values = new ArrayList<>();
        for (Operand v : valuesArg.ArrayValue()) {
            if (v.IsNumber()) {
                values.add(v.NumberValue());
            } else {
                Operand v2 = v.ToNumber(String.format("Function '%s' parameter 1 is error!", Name()));
                if (v2.IsErrorOrNone()) return v2;
                values.add(v2.NumberValue());
            }
        }

        if (values.size() == 0) {
            return ParameterError(1);
        }

        boolean hasPositive = false;
        boolean hasNegative = false;
        for (BigDecimal v : values) {
            if (v.compareTo(BigDecimal.ZERO) > 0) hasPositive = true;
            if (v.compareTo(BigDecimal.ZERO) < 0) hasNegative = true;
        }
        if (!hasPositive || !hasNegative) {
            return ParameterError(1);
        }

        BigDecimal guess = new BigDecimal("0.1");
        if (func2 != null) {
            Operand guessArg = GetNumber_2(engine, tempParameter);
            if (guessArg.IsErrorOrNone()) return guessArg;
            guess = guessArg.NumberValue();
        }

        try {
            BigDecimal irr = NewtonRaphsonIRR(values, guess);
            return Operand.Create(irr);
        } catch (Exception e) {
            return FunctionError();
        }
    }

    private BigDecimal NewtonRaphsonIRR(List<BigDecimal> values, BigDecimal guess) {
        BigDecimal rate = guess;
        for (int iter = 0; iter < 100; iter++) {
            BigDecimal npv = BigDecimal.ZERO;
            BigDecimal dnpv = BigDecimal.ZERO;

            for (int i = 0; i < values.size(); i++) {
                BigDecimal factor = BigDecimal.valueOf(Math.pow(BigDecimal.ONE.add(rate).doubleValue(), i));
                npv = npv.add(values.get(i).divide(factor, MathContext.DECIMAL128));
                dnpv = dnpv.subtract(
                        BigDecimal.valueOf(i).multiply(values.get(i))
                                .divide(factor.multiply(BigDecimal.ONE.add(rate)), MathContext.DECIMAL128));
            }

            if (dnpv.abs().compareTo(new BigDecimal("1e-12")) < 0) break;
            BigDecimal newRate = rate.subtract(npv.divide(dnpv, MathContext.DECIMAL128));

            if (newRate.subtract(rate).abs().compareTo(new BigDecimal("1e-10")) < 0) {
                return newRate;
            }
            rate = newRate;
        }
        throw new RuntimeException("IRR calculation did not converge");
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.ARRAY);
        if (func2 != null) func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
    }
}
